packages/keystone/adapterCache.js

```diff
@@ -0,0 +1,514 @@
+ *
+ *
+ * Statistics:
+ *
+ * - If cache logging is turned on, then statistics is shown on any log event..
+ *
+ * Notes:
+ *
+ * - Adapter level cache do not cache complex requests. A request is considered complex, if Where query contains other relationships
+ *  - Knex implicitly (without calling .update()) performs updates on many:true fields.
+ *    Implicit updates are present in two private knex adapter methods: _createOrUpdateField() and _setNullByValue()
+ *    It happens in case of update/create on many:true field and leads to inconsistent cache state.
+ *    We overcome this by explicitly patching those two methods to raise exception whenever they are called.
+ *    We also do not cache lists that have many: true relations or are dependant of this relations.
+ */
+
+const { get, cloneDeep, floor } = require('lodash')
+const LRUCache = require('lru-cache')
+
+const { getLogger } = require('./logging')
+const { queryHasField } = require('./queryHasField')
+const { getRedisClient } = require('./redis')
+
+const UPDATED_AT_FIELD = 'updatedAt'
+const STATE_REDIS_KEY_PREFIX = 'adapterCacheState'
+
+const logger = getLogger('adapterCache')
+
+
+class AdapterCache {
+
+    constructor ( { enabled, excludedLists, maxCacheKeys, logging } ) {
+        try {
+            this.enabled = !!enabled
+
+            // Redis is used as State:
+            // Note: Redis installation is modified with custom commands. Check getStateRedisClient for details
+            // State: { listName -> lastUpdate }
+            this.redisClient = this.getStateRedisClient()
+
+            // This mechanism allows to skip caching some lists.
+            // Useful for hotfixes or disabling cache for business critical lists
+            this.excludedLists = excludedLists || []
+
+            // This mechanism allows to control garbage collection.
+            this.maxCacheKeys = maxCacheKeys || 1000
+
+            // Cache: { listName -> queryKey -> { response, lastUpdate, score } }
+            this.cache = new LRUCache({ max: this.maxCacheKeys })
+
+            // Logging allows to get the percentage of cache hits
+            this.logging = !!logging
+
+            this.totalRequests = 0
+            this.cacheHits = 0
+
+        } catch (e) {
+            this.enabled = false
+            logger.warn(`ADAPTER_CACHE: Bad config! reason ${e}`)
+        }
+    }
+
+    /**
+     * Sometimes we might get a situation when two instances try to update state with different timestamp values.
+     * So, we might have two commands ordered to execute by internal redis queue:
+     * 1. SET some_key 1670000000010
+     * 2. SET some_key 1670000000009
+     * In result (GET some_key) will equal 1670000000009, but the correct value is 1670000000010!
+     * So to counter this we write custom redis function that will update value only if it is bigger!
+     */
+    getStateRedisClient () {
```

Comment:
Static function! it's better to extract it from this class. Or use redis definition pattern like in: RedisGuard